### PR TITLE
JDK-8269650: Optimize gc-locker in [Get|Release]StringCritical for latin string

### PR DESCRIPTION
**Pull request by Hamlin-Li**
Currently, JNI GetStringCritical locks gc locker for all strings including latin and non-latin until ReleaseStringCritical.
But for latin, it's not necessary to still lock gc locker after GetStringCritical, as it's copied anyway whether obj pining is supported or not, so it's fine to unlock gc locker after GetStringCritical.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed

### Issue
 * [JDK-8269650](https://bugs.openjdk.java.net/browse/JDK-8269650): Optimize gc-locker in [Get|Release]StringCritical for latin string

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4637/head:pull/4637` \
`$ git checkout pull/4637`

Update a local copy of the PR: \
`$ git checkout pull/4637` \
`$ git pull https://git.openjdk.java.net/jdk pull/4637/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4637`

View PR using the GUI difftool: \
`$ git pr show -t 4637`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4637.diff">https://git.openjdk.java.net/jdk/pull/4637.diff</a>

</details>

### CODE REVIEW

**Review by tschatzl — CHANGES_REQUESTED** [review]
Actually I think the *String* object can be unlocked regardless of `is_latin1` or not. The code returns the *char array* that the native code is going to process after all - which is not locked *at all* but probably should be. I filed [JDK-8269661](https://bugs.openjdk.java.net/browse/JDK-8269661) for this.

Which should probably be fixed first, because if the change correctly unconditionally unlocked the String object, Shenandoah would start to fail.
It probably works at this time because the String object and the char array are typically located in the same region anyway.

**Review by dholmes-ora — COMMENTED** [review]
Hi Hamlin,

This seems quite reasonable - so reasonable that I really need to know why we were not doing this from day one. So I need to do a bit of digging into the history here, but that will have to wait for tomorrow morning. :)

Thanks,
David

**Comment by tschatzl** [comment]
The gc team would like to ask you to wait on pushing/finishing this optimization after the fix for [JDK-8269661](https://bugs.openjdk.java.net/browse/JDK-8269661) bubbles up to JDK 18. JDK-8269661 is a P2 bug that should be fixed in JDK 17 too; it would be more convenient for us to fix it there first, then wait until we merge changes from there to 18.

Thanks.

**Comment by kimbarrett** [comment]
It turns out that fixing JDK-8269661 involves some code rearrangement that makes it both simple and obvious to not do the gc-locker/pinning in the latin1 case; it would actually make things messier to not include that change.  So unless David's research comes up with something unexpected, I'll probably take over the optimization as part of that bug fix.  Still need to do more testing before sending out a PR.

**Comment by kimbarrett** [comment]
> _Mailing list message from [David Holmes](mailto:david.holmes@oracle.com) on [hotspot-dev](mailto:hotspot-dev@mail.openjdk.java.net):_
> 
> On 30/06/2021 10:35 pm, Thomas Schatzl wrote:
> 
> > On Wed, 30 Jun 2021 11:55:49 GMT, Hamlin Li <mli at openjdk.org> wrote:
> > > Currently, JNI GetStringCritical locks gc locker for all strings including latin and non-latin until ReleaseStringCritical.
> > > But for latin, it's not necessary to still lock gc locker after GetStringCritical, as it's copied anyway whether obj pining is supported or not, so it's fine to unlock gc locker after GetStringCritical.
> > 
> > 
> > Actually I think the *String* object can be unlocked regardless of `is_latin1` or not. The code returns the *char array* that the native code is going to process after all - which is not locked *at all* but probably should be. I filed [JDK-8269661](https://bugs.openjdk.java.net/browse/JDK-8269661) for this.
> 
> I admit I do not know how String objects are laid out since compact
> strings came along but IIRC before that the char* pointed to an actual
> array embedded in the String's value, and so the String (and its value
> array) had to be pinned/locked.
> 
> David

I don’t see how that embedding of the char array could work (without bespoke GC support or something like Valhalla).
I also don’t see any sign of that in the code history.

**Pull request closed without merging**
